### PR TITLE
Implementing optional parameter `restart` to tmr:start()

### LINKED ISSUE #3020: tmr:start() does not restart started timer
### Expected behavior
https://nodemcu.readthedocs.io/en/master/modules/tmr/#tobjstart
says that `tmr:start()` **restarts** timer. It's not true:

### Actual behavior
`tmr:start()` will **restart** timer (start it count from zero)

### Test code
```Lua
> t = tmr.create()
> t:register( 3000, tmr.ALARM_SEMI, function( t ) print( 'alarm' ) end)
> t:start()
> -- wait 2 seconds
> t:start()
> -- alarm message will be printed 3 seconds after the first t:start() call, not the second one as expected.
```

### Workaround
```Lua
t:stop()
t:start()
```

### NodeMCU startup banner
NodeMCU 3.0.1.0, build type: integer
modules: file,gpio,net,node,rtctime,sntp,tmr,uart,wifi
build 2020-01-27 17:39 powered by Lua 5.1.4 on SDK 3.0.2(824dc80)

### Hardware
ESP8266 wemos D1 (not mini)

### PR DESCRIPTION
**Pull request by vsky279**
Fixes #3020.

- [x] This PR is for the `dev` branch rather than for `master`.
- [x] This PR is compliant with the [other contributing guidelines](https://github.com/nodemcu/nodemcu-firmware/blob/dev/CONTRIBUTING.md) as well (if not, please describe why).
- [x] I have thoroughly tested my contribution.
- [x] The code changes are reflected in the documentation at `docs/*`.

Implements the solution discussed in #3020 as wrapped up by @HHHartmann 

Also removes unecessary check of lua_newuserdata result (also in ads1115 module).

### CODE REVIEW

File: app/modules/ads1115.c [hunk 1]
```diff
@@ -252,9 +252,6 @@ static int ads1115_lua_register(lua_State *L, uint8_t chip_id) {
         return luaL_error(L, "unexpected config value (%p) please reset device before calling this function", config_read);
     }
     ads_ctrl_ud_t *ads_ctrl = (ads_ctrl_ud_t *)lua_newuserdata(L, sizeof(ads_ctrl_ud_t));
-    if (NULL == ads_ctrl) {
-        return luaL_error(L, "ads1115 malloc: out of memory");
-    }
```
Comment (nwf) [inline app/modules/ads1115.c:257]: Unrelated change?  (Though I think one we should take, but not part of this PR.)

File: app/modules/tmr.c [hunk 2]
```diff
@@ -184,16 +184,21 @@ static int tmr_register(lua_State* L){
 // Lua: tmr.start( id / ref )
```
Comment (nwf) [inline app/modules/tmr.c:184]: `Lua: tmr.start(ref, [restart])`?  More generally, if you're wanting to tidy `tmr`, could you drop the `id` from all of the `Lua:` comments?

File: app/modules/tmr.c [hunk 3]
```diff
@@ -363,7 +368,6 @@ static int tmr_softwd( lua_State* L ){
 // Lua: tmr.create()
 static int tmr_create( lua_State *L ) {
 	tmr_t ud = (tmr_t)lua_newuserdata(L, sizeof(timer_struct_t));
-	if (!ud) return luaL_error(L, "not enough memory");
```
Comment (nwf) [inline app/modules/tmr.c:366]: Unrelated, but we should take it as it's part of `tmr`.

File: app/modules/ads1115.c (hunk 1 above)
Comment (vsky279) [reply app/modules/ads1115.c:257]: Ok. Let's make it another PR.

File: app/modules/tmr.c (hunk 2 above)
Comment (vsky279) [reply app/modules/tmr.c:184]: Ok, let's do it.

**Comment by TerryE** [comment]
Lukas, I tidied up ``tmr.c` (= quite a lot of rework) for my #3078.  If you don't mind, it's going to be a _lot_ easier to add in this couple of changes rather than having to do a three-way merge. Thanks.

**Comment by vsky279** [comment]
Of course not. So you will include these few lines in your commit?

File: app/modules/tmr.c [hunk 4]
```diff
@@ -125,17 +125,24 @@ static int tmr_register(lua_State* L) {
...
 }
 
-// Lua: t:start()
+// Lua: t:start( [restart] )
 static int tmr_start(lua_State* L){
 	tmr_t *tmr = (tmr_t *) luaL_checkudata(L, 1, "tmr.timer");
-  int idle = tmr->mode & TIMER_IDLE_FLAG;
+	lua_settop(L, 2);
+	luaL_argcheck(L, lua_isboolean(L, 2) || lua_isnil(L, 2), 2, "boolean expected");
+	int restart = lua_isboolean(L, 2) ? lua_toboolean(L, 2) : 0;
```
Comment (marcelstoer) [inline app/modules/tmr.c:133]:
```suggestion
	int restart = lua_isboolean(L, 2);
```
You already verified it's a boolean in the line above.

File: app/modules/tmr.c [hunk 5]
```diff
@@ -125,17 +125,24 @@ static int tmr_register(lua_State* L) {
...
-  	tmr->mode &= ~TIMER_IDLE_FLAG;
+	//we return false if the timer is not idle
+	int idle = tmr->mode&TIMER_IDLE_FLAG;
+	if(!(idle || restart)){
+		lua_pushboolean(L, 0);
+	}else{
+		if (!idle) {os_timer_disarm(&tmr->os);}
+		tmr->mode &= ~TIMER_IDLE_FLAG;
 		os_timer_arm(&tmr->os, tmr->interval, tmr->mode==TIMER_MODE_AUTO);
   }
```
Comment (marcelstoer) [inline app/modules/tmr.c:147]:
```suggestion
	}
```

Comment (TerryE) [reply app/modules/tmr.c:147]: This is an example where some of `tmr.c` uses `\t` and some uses `  ` (two spaces) as the logical indent.  But if we aren't going to fix the whole mix as Marcel suggests, then we should at least be consistent within each function.

File: app/modules/tmr.c (hunk 4 above)
Comment (vsky279) [reply app/modules/tmr.c:133]: But it can still be `nil` if default parameter is used. I have not checked that `lua_toboolean(L, 2)` returns 0 if parameter is `nil`. So 👍

File: app/modules/tmr.c (hunk 5 above)
Comment (vsky279) [reply app/modules/tmr.c:147]: There were more to be switched to tab. I prefer spaces but tab prevails. So I switched them to tab.

**Comment by vsky279** [comment]
Oups. A bit of fight with @marcelstoer in commits. I hope the result is fine.

**Comment by marcelstoer** [comment]
> I hope the result is fine.

As #1168 is now closed we maybe shouldn't argue about it but replacing spaces with tabs is a move in the wrong direction IMO.

I'm happy to merge this though.

**Comment by vsky279** [comment]
spaces -> tabs - I only made it consistent with the rest of the file where tabs are used.
In the following PR we can replace tabs to spaces for good.
